### DISCUSSION TITLE
what to do if MDS are failing

### DISCUSSION THREAD

**Discussion by ahmedTouati** [rook/rook#8469] [Q&A]
Hello, we had to destroy cephcluster with all data in it multiple times because of MDS failing without any clear message , i had to repair once by following ceph repair mds guide but that's take time and is complicated. I think there should be something clear to do when MDS are failling :

```
"MDS_ALL_DOWN":{"HEALTH_ERR" "1 filesystem is offline"} "MDS_DAMAGE":{"HEALTH_ERR" "1 mds daemon damaged"}
```

**Comment by travisn** [comment]
1. Are you able to repro this? Or it was just a one-time failure related to the mds?
2. Do you have logs from the rook operator or mds pod?
3. No other health errors or warnings in `ceph status`?

**Reply by ahmedTouati** [reply]
@travisn the exception is not clear after i checked rook-operator logs, but i had to restore the whole cluster from last backup, any way not it's working

**Reply by osaffer** [reply]
Our infrastructure is running on VMWare. We had problem with datastore. 
Then we had the same error : 
MDS_ALL_DOWN: 1 filesystem is offline
MDS_DAMAGE: 1 mds daemon damaged
FS_DEGRADED: 1 filesystem is degraded

the MDS were not able to start, in logs it was showing starting. 
So I tried to find more details inside the mds pod: 

k exec -it rook-ceph-mds-postgresfs-a-54fb59c584-8zv8p -- sh
Defaulted container "mds" out of: mds, chown-container-data-dir (init)
			  
			  
ceph-mds --id=postgresfs-a --debug_mds 1 --no-mon-config
2021-09-24T13:18:53.718+0000 7fdc3ca9c600 -1 asok(0x5560b22dc000) AdminSocketConfigObs::init: failed: AdminSocket::bind_and_listen: failed to bind the UNIX domain socket to '/var/run/ceph/ceph-mds.postgresfs-a.asok': (17) File exists
2021-09-24T13:18:53.718+0000 7fdc3ca9c600 -1 auth: unable to find a keyring on /var/lib/ceph/mds/ceph-postgresfs-a/keyring: (2) No such file or directory
2021-09-24T13:18:53.718+0000 7fdc3ca9c600 -1 AuthRegistry(0x5560b2ebc140) no keyring found at /var/lib/ceph/mds/ceph-postgresfs-a/keyring, disabling cephx

But when I checked  In the deployment of mds-postgresfs-a

 - name: rook-ceph-mds-postgresfs-a-keyring
              readOnly: true
              mountPath: /etc/ceph/keyring-store/


And the command is good : 

 command:

            - ceph-mds

          args:

            - '--fsid=23df4bc9-b147-4849-9c23-66c6010ec1b6'

            - '--keyring=/etc/ceph/keyring-store/keyring'

            - '--log-to-stderr=true'

            - '--err-to-stderr=true'

            - '--mon-cluster-log-to-stderr=true'

            - '--log-stderr-prefix=debug '

            - '--default-log-to-file=false'

            - '--default-mon-cluster-log-to-file=false'

            - '--mon-host=$(ROOK_CEPH_MON_HOST)'

            - '--mon-initial-members=$(ROOK_CEPH_MON_INITIAL_MEMBERS)'

            - '--id=postgresfs-a'

            - '--setuser=ceph'

            - '--setgroup=ceph'

            - '--foreground'




Despite a restore of the VM and disks, I have not been able to come back to normal situation.

How to recreate the MDS ?

**Reply by osaffer** [reply]
in crash collector I have : 

WARNING:ceph-crash:post /var/lib/ceph/crash/2021-09-23T12:16:56.028537Z_17b5ea51-5ee8-4208-83f1-3c1e702934ac as client.admin failed: b'[errno 13] RADOS permission denied (error connecting to the cluster)\n'
WARNING:ceph-crash:post /var/lib/ceph/crash/2021-09-23T12:16:56.028537Z_17b5ea51-5ee8-4208-83f1-3c1e702934ac as client.crash.rook-ceph-crashcollector-adakub2.do.gov.lu-5f8595d894-mp6rm failed: b'[errno 13] RADOS permission denied (error connecting to the cluster)\n

**Comment by BlaineEXE** [comment]
The MDSes are stateless (they don't store data on the host), so I think you can just delete the MDS deployment and restart the Ceph operator to have it be recreated. It sounds like there could be more issues with the cluster, however, if the RADOS permissions are denied. Are the OSDs failing?

**Reply by osaffer** [reply]
Hi,

Thanks for your reply.

I had already restarted the ceph operator, but it did not work.

ALL OSD are up

**Comment by reevejd** [comment]
I ran into an `MDS_DAMAGE` issue today. It doesn't seem to be the exact same issue the OP had (no rados permission issues in my case), but the symptoms are the same and it matches the topic title. Take this with a grain of salt because I'm not a ceph expert, but for me the fix was: 

1. Identify the damaged MDS deployments by running:          
       ``` 
        kubectl exec deploy/rook-ceph-tools -- /bin/sh -c 'ceph tell mds.* damage ls 
       ```
       
     For some of the MDSes I saw this error:

        Error ENOSYS: 
        command terminated with exit code 38

1. Delete the corresponding MDS deployments which showed an error in step 1.   The deployments weren’t recreated automatically at this point, but that was expected because the cluster was in an error state and the operator won't do some operations when the cluster is unhealthy.   Confirm that there is no remaining damage by running this again: 
       ``` 
        kubectl exec deploy/rook-ceph-tools -- /bin/sh -c 'ceph tell mds.* damage ls 
       ```

1. In order to get the cluster back into a healthy state, run:
    ```
    kubectl exec deploy/rook-ceph-tools -- /bin/sh -c 'ceph mds repaired 0'
    ```

After those steps the operator recreated the deleted MDS deployment and the cluster health quickly recovered.

Is this something the operator could automatically handle? I don't know enough about rook/ceph to predict in which cases the MDS deployments being deleted would be sufficient to fix the problem.
